File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/persistence/CachedProductRepository.java
package com.zappymart.adapters.persistence;

import com.zappymart.application.ports.ProductRepository;
import com.zappymart.domain.catalogue.Product;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;

public final class CachedProductRepository implements ProductRepository {

    private final ProductRepository catalogueBehindTheCache;
    private final AtomicReference<List<Product>> remembered = new AtomicReference<>();

    public CachedProductRepository(ProductRepository catalogueBehindTheCache) {
        this.catalogueBehindTheCache = catalogueBehindTheCache;
    }

    @Override
    public List<Product> catalogue() {
        List<Product> known = remembered.get();
        if (known != null) {
            return known;
        }
        List<Product> read = catalogueBehindTheCache.catalogue();
        remembered.set(read);
        return read;
    }

    @Override
    public Optional<Product> byId(String productId) {
        return catalogue().stream().filter(product -> product.id().equals(productId)).findFirst();
    }

    @Override
    public Optional<Product> bySlug(String slug) {
        return catalogue().stream().filter(product -> product.slug().equals(slug)).findFirst();
    }

    @Override
    public List<Product> byIds(Collection<String> productIds) {
        return catalogue().stream().filter(product -> productIds.contains(product.id())).toList();
    }

    @Override
    public void reduceStock(String productId, int quantity) {
        catalogueBehindTheCache.reduceStock(productId, quantity);
        forget();
    }

    public void forget() {
        remembered.set(null);
    }
}

File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/persistence/CachedProductRepository.kt
package nl.zappymart.adapters.persistence

import java.util.concurrent.ConcurrentHashMap
import nl.zappymart.application.ports.ProductRepository
import nl.zappymart.domain.catalogue.Product
import nl.zappymart.domain.catalogue.ProductSpecification

class CachedProductRepository(private val products: ProductRepository) : ProductRepository {

    private data class PageKey(val specification: ProductSpecification, val size: Int, val afterProductId: String?)

    private val pages = ConcurrentHashMap<PageKey, List<Product>>()

    private val counts = ConcurrentHashMap<ProductSpecification, Int>()

    override fun page(specification: ProductSpecification, size: Int, afterProductId: String?): List<Product> =
        pages.computeIfAbsent(PageKey(specification, size, afterProductId)) { key ->
            products.page(key.specification, key.size, key.afterProductId)
        }

    override fun count(specification: ProductSpecification): Int =
        counts.computeIfAbsent(specification) { key -> products.count(key) }

    override fun findById(productId: String): Product? = products.findById(productId)

    override fun findBySlug(slug: String): Product? = products.findBySlug(slug)

    override fun findAllByIds(productIds: List<String>): List<Product> = products.findAllByIds(productIds)

    override fun reduceStock(quantityPerProductId: Map<String, Int>) {
        products.reduceStock(quantityPerProductId)
        forget()
    }

    fun forget() {
        pages.clear()
        counts.clear()
    }
}
